build.assets/tooling/cmd/changelog/prparser.go

```diff
@@ -0,0 +1,188 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	// timeNow is a convenience variable to signal that search should include PRs up to current time
+	timeNow = ""
+)
+
+type parsedPR struct {
+	// Summary is the changelog summary extracted from a PR
+	Summary string
+	Number  int
+	URL     string
+}
+
+const (
+	ossCLTemplate = `
+{{- range . -}}
+* {{.Summary}} [#{{.Number}}]({{.URL}})
+{{ end -}}
+`
+	entCLTemplate = `
+{{- range . -}}
+* {{.Summary}}
+{{ end -}}
+`
+)
+
+var (
+	// clPattern will match a changelog format with the summary as a subgroup.
+	// e.g. will match a line "changelog: this is a changelog" with subgroup "this is a changelog".
+	clPattern = regexp.MustCompile(`[Cc]hangelog: +(.*)`)
+
+	ossCLParsedTmpl = template.Must(template.New("oss cl").Parse(ossCLTemplate))
+	entCLParsedTmpl = template.Must(template.New("enterprise cl").Parse(entCLTemplate))
+)
+
+// pr is the expected output format of our search query
+type pr struct {
+	Body   string `json:"body,omitempty"`
+	Number int    `json:"number,omitempty"`
+	Title  string `json:"title,omitempty"`
+	URL    string `json:"url,omitempty"`
+}
+
+type changelogGenerator struct {
+	isEnt bool
+	dir   string
+}
+
+// dateRangeFormat takes in a date range and will format it for GitHub search syntax.
+// to can be empty and the format will be to search everything after from
+func dateRangeFormat(from, to string) string {
+	if to == "" {
+		return fmt.Sprintf(">%s", from)
+	}
+	return fmt.Sprintf("%s..%s", from, to)
+}
+
+// findChangelog will parse a body of a PR to find a changelog.
+func findChangelog(commentBody string) (found bool, summary string) {
+	// If a match is found then we should get a non empty slice
+	// 0 index will be the whole match including "changelog: *"
+	// 1 index will be the subgroup match which does not include "changelog: "
+	m := clPattern.FindStringSubmatch(commentBody)
+	if len(m) > 1 {
+		return true, m[1]
+	}
+	return false, ""
+}
+
+func prettierSummary(cl string) string {
+	cl = strings.TrimSpace(cl)
+	if !strings.HasSuffix(cl, ".") {
+		cl += "."
+	}
+	return cl
+}
+
+// ghListPullRequests is a wrapper around the `gh` command to list PRs
+// searchQuery should follow the GitHub search syntax
+func (c *changelogGenerator) ghListPullRequests(searchQuery string) (string, error) {
```

Comment:
What do you think about factoring this logic out into a mini "github" package, e.g. alongside your "git" package? Then you would import it here and go `github.ListPullRequests(xxx)` and the Github-specific logic would be encapsulated there which would make code easier to comprehend too (IMO).

Reply:
>  mini "github" package

This is another reason for moving this to shared-workflows - [there is already a `github` package there that we maintain](https://github.com/gravitational/shared-workflows/blob/main/bot/internal/github/github.go). Same thing with git as well I believe.

Reply:
I did think about it but I figured to keep the solution simple I'd just use the `gh` CLI. The `gh` CLI handles pagination so I don't have to! Although to be fair I don't think we'll realistically need to actually paginate.

I was thinking a more robust solution would be based on [Search Issues And Pull Requests](https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#search-issues-and-pull-requests) e.g. `github.SearchPullRequests(xxx)`. I might try my hand at it since it does already exist.